### ISSUE TITLE
Coinbase Pro api sunsetting

### ISSUE THREAD

**Issue by nibarulabs** [knowm/XChange#4514]
The Pro features will be moving to Advanced Trade.

https://blog.coinbase.com/hello-advanced-trade-goodbye-coinbase-pro-5b0715b03ef0

Our product uses Pro and we'll need to do some work to make sure that the regular Coinbase adapter works with Advanced Trade. I'm taking a quick look through their updated api docs.

One thing I don't see in the updated (current) docs - https://docs.cloud.coinbase.com/exchange/reference/exchangerestapi_getaccounts - is any websocket info (I guess it says Rest Api at the top, so I'm assuming some ws section will appear later?).

There is ws info on the guides page - https://docs.cloud.coinbase.com/exchange/docs/websocket-overview - so maybe it'll all just magically work with existing Pro code.

We have a decent amount of Coinbase Pro users, so we will have to have an updated working adapter. I'll probably take this on in September'ish, which should be plenty of time before end of year. I anticipate the update shouldn't take too long - weeks? days?

I think this will probably figure itself out, but we have a few things to resolve:

- Fold all the Advanced Trade into the Coinbase adapter code and call it good
- Delete (I guess at a minimum deprecate CB Pro adapter with some console warnings of sunset and removal) CB Pro at some point

If anyone starts work on this before then, can you reference this issue just to keep things in some order? I'll update comments and the PR code when it's ready.

Comments/thoughts welcomed.

**Comment by Eddman** [comment]
Any news on this one? Some account are already being disabled on CoinbasePro and users forces to use the new AdvanceTrading.

**Comment by nibarulabs** [comment]
Hello, I am starting to take a look at this today. Not sure how long it will take. I'll post updates as I have them.

**Comment by nibarulabs** [comment]
Update. I went through all the existing CB Pro code and the docs for the new Advanced api.

Good news is Coinbase cleaned a lot of things up; the old api was a little rickety. And I can definitely do all the work (I'm capable) and I just started a new module for Advanced.

Bad (sorta) news. I don't think any/much of the old Pro module code is reusable. I'll take what is though. Again, it's not a difficult project, just time. If I was working on this full time, I could prob knock it out in 2 weeks or less. However, since we're in free contribution and day job time :) it'll take longer.

There's an opportunity to test more and I'll have a lot more testing in than Pro had. I also have a product that will be able to test the library so once the code is in a good spot, I will start interfacing and testing in the real world.

Oh, and this is just the exchange code, I know the stream code has to be done (my product will need that as well) and I'm not sure how long that will take. I will prob do the stream code in another PR. I'll update as needed :+1: .

**Comment by timmolter** [comment]
I'd say leave the coinbase-pro module how it is and add a completely new one called coinbase-advanced, and implement the API there as much as you can. Then later when coinbase-pro really gets shuts off, we can delete that module.

**Comment by nibarulabs** [comment]
Yah, I already figured coinbase-pro would need to exist as is and have been creating a new advanced module.

Sounds good, thanks for the quick reply. That lines up with what I was thinking.

**Comment by Eddman** [comment]
Isn't the Advanced API just the unimplemented part of [xchange-coinbase](https://github.com/knowm/XChange/tree/develop/xchange-coinbase) module? I thought that Advanced API just adds trading API using the new v3 endpoints to standard Coinbase v2 API....so I'd expect to just implement all unimplemented methods in this module.

**Comment by nibarulabs** [comment]
Advanced has a completely different set of uri endpoints. I looked for other examples in this repo to see if any other modules were using two different ssl url properties and none were - I was thinking if so, I'd just add it to the coinbase module.

**Comment by nibarulabs** [comment]
@Eddman Ok, you're right. The xchange-coinbase module has the ssl uri https://coinbase.com. For the Advanced functionality to be folded in, that uri would have to change to https://api.coinbase.com. The Coinbase interface paths would all need to be updated as well.  

Currently the path for the Coinbase uri is set to ```api/v1``` and it will need to be removed and each annotation would need to include ```api/v2``` for the existing endpoints, ```api/v3``` for advanced endpoints. I'll look through the docs to make sure that the existing endpoints are valid.

Thanks for pointing that out. This will obviously make things cleaner and less time to implement.

**Comment by nibarulabs** [comment]
Providing an update here.

I'm putting my effort here on pause since my work is blocked by Coinbase. I was working through fetching the candles for a pair. They have this endpoint now behind an auth wall (ok, np) and while I can make requests to other authenticated endpoints, fetching candles always responds with unauthorized 401. My api credentials are correct and I even repeated the issue with some Node code.

A co-worker got their support on the phone and we raised the issue. It got filed as we're looking into it. I went on the forums and again, no solution - I couldn't get anyone from their support to respond.

At this point my takeaway is the advanced api is not ready for prime time. Seems like there are a significant amount of others reporting basic problems in the forums. Coinbase announced this venture last year. My experience working on development teams is we wouldn't announce a major service sunset without at least being in beta.

I'm just leaving this here for someone that might pick this up later. I don't have anything worthwhile to commit and I'll just leave things as is. Cheers.
